haystack/document_stores/filter_utils.py

```diff
@@ -187,107 +205,265 @@ def convert_to_elasticsearch(self):
         """
         pass
 
+    @abstractmethod
+    def convert_to_weaviate(self):
+        """
+        Converts the ComparisonOperation instance to a Weaviate comparison operator.
+        """
+        pass
+
+    @abstractmethod
+    def invert(self) -> "ComparisonOperation":
+        """
+        Inverts the ComparisonOperation.
+        Necessary for Weaviate as Weaviate doesn't seem to support the 'Not' operator anymore.
+        (https://github.com/semi-technologies/weaviate/issues/1717)
+        """
+        pass
+
+    def _get_weaviate_datatype(
+        self, value: Optional[Union[str, int, float, bool]] = None
+    ) -> Tuple[str, Union[str, int, float, bool]]:
+        """
+        Determines the type of the comparison value and converts it to RFC3339 format if it is as date,
+        as Weaviate requires dates to be in RFC3339 format including the time and timezone.
+
+        """
+        if value is None:
+            assert not isinstance(self.comparison_value, list)  # Necessary for mypy
+            value = self.comparison_value
+
+        if isinstance(value, str):
+            # Check if comparison value is a date
+            try:
+                value = convert_date_to_rfc3339(value)
+                data_type = "valueDate"
+            # Comparison value is a plain string
+            except ValueError:
+                data_type = "valueString"
+        elif isinstance(value, int):
+            data_type = "valueInt"
+        elif isinstance(value, float):
+            data_type = "valueNumber"
+        elif isinstance(value, bool):
+            data_type = "valueBoolean"
+        else:
+            raise ValueError(
+                f"Unsupported data type of comparison value for {self.__class__.__name__}."
+                f"Value needs to be of type str, int, float, or bool."
+            )
+
+        return data_type, value
+
 
 class NotOperation(LogicalFilterClause):
     """
     Handles conversion of logical 'NOT' operations.
     """
 
-    def convert_to_elasticsearch(self):
+    def convert_to_elasticsearch(self) -> Dict[str, Dict]:
         conditions = [condition.convert_to_elasticsearch() for condition in self.conditions]
         conditions = self._merge_es_range_queries(conditions)
         return {"bool": {"must_not": conditions}}
 
+    def convert_to_weaviate(self) -> Dict[str, Union[str, int, float, bool, List[Dict]]]:
+        conditions = [condition.invert().convert_to_weaviate() for condition in self.conditions]
+        if len(conditions) > 1:
+            return {"operator": "And", "operands": conditions}
+        else:
+            return conditions[0]
+
+    def invert(self) -> "OrOperation":
+        return OrOperation([condition.invert() for condition in self.conditions])
```

Comment:
The `invert` methods for `AndOperation` and `OrOperation` I understood easily and it's a great use case for De Morgan's laws. 👍 However, at first I found it difficult to understand why the `invert` of the NotOperation is an "OrOperation". You could add a comment here explaining that conditions in the list `self.conditions` are by default combined with AND and therefore it becomes an OR.

Reply:
I added a comment there. Returning an `OrOperation` was actually wrong here; this should be done in the `convert_to_weaviate` method. The `invert` method of the `NotOperation` is in fact only called for nested Not operations.